singularity-eos/eos/eos_vinet.hpp

```diff
@@ -0,0 +1,366 @@
+  PORTABLE_INLINE_FUNCTION void
+  DensityEnergyFromPressureTemperature(const Real press, const Real temp, Real *lambda,
+                                       Real &rho, Real &sie) const;
+  inline void Finalize() {}
+  static std::string EosType() { return std::string("Vinet"); }
+  static std::string EosPyType() { return EosType(); }
+
+ private:
+  //AEM: I am not sure what this does so leave it in.
+  static constexpr const unsigned long _preferred_input =
+      thermalqs::density | thermalqs::temperature;
+  Real _rho0, _T0, _B0, _BP0, _A0, _Cv0, _E0, _S0;
+  const Real *_expconsts;
+  Real _VIP[43], _d2to40[39];
+  PORTABLE_INLINE_FUNCTION void CheckVinet();
+  PORTABLE_INLINE_FUNCTION void Vinet_F_DT_func(const Real rho, const Real T, 
+		  				Real *output) const;
+};
+
+PORTABLE_INLINE_FUNCTION void Vinet::CheckVinet() {
+
+  int ind;
+
+  if (_rho0 < 0.0){
+    printf("testing rho0\n");
+    EOS_ERROR("Required Vinet model parameter rho0 not set");
+    return;
+  }
+  if (_T0 < 0.0){
+    printf("testing T0\n");
+    EOS_ERROR("Required Vinet model parameter T0 not set");
+    return;
+  }
+  if (_B0 < 0.0){
+    printf("testing B0\n");
+    EOS_ERROR("Required Vinet model parameter B0 not set");
+    return;
+  }
+  if (_BP0 < 1.0){
+    printf("testing BP0\n");
+    EOS_ERROR("Required Vinet model parameter BP0 not set or too small (<1)");
+    return;
+  }
+  if (_A0 < 0.0){
+    printf("testing A0\n");
+    EOS_ERROR("Required Vinet model parameter A0 not set");
+    return;
+  }
+  if (_Cv0 < 0.0){
+    printf("testing Cv0\n");
+    EOS_ERROR("Required Vinet model parameter Cv0 not set");
+    return;
+  }
+
+
+  for(ind = 0; ind < 39; ind++) _d2to40[ind]=_expconsts[ind];
+
+  _VIP[0] = _B0/_rho0 + pow(_A0*_B0/_rho0,2.0)*_T0/_Cv0;      // sound speed squared
+  _VIP[1] = 3.0/2.0*(_BP0-1.0); // exponent
+  _VIP[2] = 1.0; // prefactor f0
+  _VIP[3] = 0.0; // prefactor f1
+
+  for (ind = 4; ind < 43; ind++) _VIP[ind] = _d2to40[ind-4];
+  ind = 42;
+  while (_VIP[ind] == 0.0) ind--;
+  while (ind >= 4) {
+    ind--;
+    _VIP[ind] = _VIP[ind] - (ind -1)/_VIP[1]*_VIP[ind+1];
+  }
+}
+
+
+PORTABLE_INLINE_FUNCTION void Vinet::Vinet_F_DT_func(const Real rho, const Real T, 
+							Real *output) const {
+  int pref0vp = -2, pref0vip = 2, ind;
+  Real x, x2inv, onemx, etatimes1mx, expetatimes1mx, temp;
+  Real sumP=0.0, sumB=0.0, sumE=0.0;
+  Real entropy, energy, pressure, dpdrho, dpdt, dedt, dedrho, soundspeed;
+
+  ind = 40;
+  while (_VIP[pref0vip+ind] == 0.0) ind--;
+
+  x = pow(_rho0/rho,1.0/3.0);  /*rho-dependent*/
+  x2inv = 1.0/pow(x,2.0);
+  onemx = 1.0-x;
+  etatimes1mx = _VIP[1]*onemx;
+  expetatimes1mx = exp(etatimes1mx);
+
+  while (ind >= 2){
+    sumP = _d2to40[pref0vp+ind]+onemx*sumP;
+    sumB = _d2to40[pref0vp+ind]*ind+onemx*sumB;
+    sumE = _VIP[pref0vip+ind]+onemx*sumE;
+    ind--;
+  }
+  while (ind >= 0){
+    sumE = _VIP[pref0vip+ind]+onemx*sumE;
+    ind--;
+  }
```

Comment:
Again, I think it would be more clear to use `for` loops rather than `while` loops.